### ISSUE TITLE
expected_version doesn't prevent publishing an event in the incorrect order

### ISSUE THREAD

**Issue by vaihtovirta** [RailsEventStore/rails_event_store#477]
Hello!

Recently I started the integration of `rails_event_store` in a couple of projects, and I run into an issue with the behavior of `expected_version` option.

## Setup (steps to reproduce)
```ruby
client = RailsEventStore::Client.new
OrderUpdated = Class.new(RailsEventStore::Event)
stream_name = "Order_1"
RailsEventStoreActiveRecord::EventInStream.where(stream: stream_name)
=> # []

event_1 = OrderUpdated.new(data: {})
client.publish(event1, stream_name: stream_name, expected_version: -1)

event_2 = OrderUpdated.new(data: {})
client.publish(event2, stream_name: stream_name, expected_version: 2)
```
## Expected behavior

- `RubyEventStore::WrongExpectedEventVersion` exception is raised since there is no event with a `position: 2` in the stream
- The stream contains the first event with `position: 0`

## Actual behavior

- `event_2` got published to the stream
- The stream contains two events with `position: 0` and  `position: 3`

```ruby
RailsEventStoreActiveRecord::EventInStream.where(stream: stream_name)
=> [#<RailsEventStoreActiveRecord::EventInStream:0x0000555950a07c70
  id: 1,
  stream: "Order_1",
  position: 0,
  event_id: "ce08208a-ae2b-4de9-932f-95cef20ac9d8",
  created_at: Wed, 24 Oct 2018 11:03:43 CEST +02:00>,
 #<RailsEventStoreActiveRecord::EventInStream:0x0000555950a04908
  id: 3,
  stream: "Order_1",
  position: 3,
  event_id: "7d136865-57de-4c01-bcbf-2f0bc199f691",
  created_at: Wed, 24 Oct 2018 11:03:46 CEST +02:00>]
```

Is it supposed behavior of `expected_version`?

That's how I understood it should work from the [docs](https://railseventstore.org/docs/expected_version/#explicit-number-integer-from-1), but maybe I'm missing something.

## Possible solution

It can be either a change in default `RailsEventStoreActiveRecord::EventRepository` or a new event repository class with the following check:

```ruby
def assert_version!
  return if expected_version == last_position

  raise RubyEventStore::WrongExpectedEventVersion
end

def last_position
  @last_position ||= RailsEventStoreActiveRecord::EventInStream.
    where(stream: stream_name).order("position DESC").first&.position || -1
end
```

Please advise if the solution is needed to achieve the expected behavior or maybe it's possible to maintain the strict publishing order with existing gem functionality. Thank you.

**Comment by pawelpacana** [comment]
Thank you for a detailed report! 

I can confirm such behaviour. That is an overlooked case rather than an intended outcome. The main driver for integer as `expected_version` was `aggregate_root` and eliminating possible race conditions. What we do there in regard to position is:
- remembering last read position (version) of a stream https://github.com/RailsEventStore/rails_event_store/blob/master/aggregate_root/lib/aggregate_root.rb#L39-L41
- assuming this position still holds true on write https://github.com/RailsEventStore/rails_event_store/blob/master/aggregate_root/lib/aggregate_root.rb#L48 and failing otherwise on database constraint (as the position is already taken)

The situation where you can append to stream with `expected_version` > last stream position is problematic since:
- `aggregate_root` expects no gaps in position https://github.com/RailsEventStore/rails_event_store/blob/master/aggregate_root/lib/aggregate_root.rb#L49 (but it would not happen from typical usage of aggregate_root that is by only aggregate_root writing to that particular stream)
- you can mess the order of events in a stream, as we sort by position on read and you can append further event with earlier position when there's a gap

```ruby
client = RailsEventStore::Client.new
DummyEvent = Class.new(RailsEventStore::Event)

client.publish(DummyEvent.new, stream_name: "dummy", expected_version: -1)
client.publish(DummyEvent.new, stream_name: "dummy", expected_version: 10)
client.publish(DummyEvent.new, stream_name: "dummy", expected_version: 5)
```

Regarding the proposed solution, the question is how this check would behave with concurrent writers — without explicit lock for reading last stream position and appending to a stream, another writer might have already succeeded. We'll have to contemplate about it more.

Related: https://eventstore.org/docs/dotnet-api/optimistic-concurrency-and-idempotence/index.html?q=optimist#idempotence

Btw. I'm curious about your usecase for `expected_version` and how it led you to aforementioned situation — I might be missing something obvious.

**Comment by paneq** [comment]
@vaihtovirta @pawelpacana - I can confirm this is a known limitation of the current implementation which uses uniq db index to detect conflicts. You might call it a bug.

It was built with the assumption that you read events, add +1 to compute next position, write next events. Exactly like aggregate_root gem does. My assumption was that it's good enough and I don't expect to see many errors in real life due to this limitation.

Detecting gaps would require probably at least one more SQL query which would reduce performance on every write so I wanted to avoid it very much. It would be easier in a 3-table-layout where we would have a separate table for streams keeping the position of last written event in a stream.

This proposed solution should be fine in terms of concurrent writes and potential race conditions because:

* if you write N+1 th event, you can safely assume you should be able to see N-th event in your transaction
* we would still have uniq db index to prevent writing on same position

> without explicit lock for reading last stream position and appending to a stream, another writer might have already succeeded.

Indeed. If we see a smaller number, we would know there is a gap and `WrongExpectedEventVersion` makes sense. If we see a bigger number, we would earlier detect a concurrency issue and rise another appropriate exception (I forgot its name). So all should be fine IMHO.

I don't think we would need a separate Repository for this. We could add a configuration option to a repository constructor such as (`verify_no_gaps: true`) for those who prefer to be on the safe side, instead of on the fast side :) 

Of course this would only make sense for writes where the strategy ( https://railseventstore.org/docs/expected_version/ ) is an explicit number. For `:any` and `:auto` there would be no change. `any` does not have deterministic write order and `auto` computes this number itself.

> Btw. I'm curious about your usecase for expected_version and how it led you to aforementioned situation — I might be missing something obvious.

Same here.

**Comment by vaihtovirta** [comment]
@pawelpacana @paneq Thank you for such quick response!

Here's a high-level overview of the use-case and setup in our projects:
- We have an external producer of events, which are received by Rails application via Amazon SQS
- The external producer sets up `expected_version` on his end
- The way SQS queue is set up now, it cannot guarantee the same order for events, so it goes like First-In Somewhen-Out :)
    - The producer publishes events in the strict order `A, B, C, D`
    - Rails application may receive this batch as `D, B, A, C`
- When events get reached to Rails app, we do some data normalization and then finally publish the event via `RailsEventStore::Client#publish`

Even the order of events may be random; each event still has a version set. 
Thus, It's crucial for us to have strict publishing order in `rails_event_store`, even if it's less performant. 
Also, it's totally fine for us to fail if there is some inconsistency in versions and retry publishing later.

**Comment by paneq** [comment]
@vaihtovirta Let me rephrase you to make sure I understand correctly

* You get an event in random order (say B)
* You try to publish B, and it should fail
* That would schedule the retry of publication of event B in the future by some kind of mechanism
* You get an event in random order (say A)
* You try to publish A, and it should succeed
* Hopefully at this moment you get B again, you try to publish B, and it should succeed

Is that what you are trying to achieve?

If so, I wonder... Technically that can work, but with increasing number of events delivered out of order, I believe, the chance of publishing an incorrect on grows more than linearly. If you have 10 events in wrong order, then it might be that you retry 9 of them before publishing the right one. That can be problematic. Perhaps you are doing something smarter by in _data normalization_ step that you mentioned.

Also, I don't think this check (for no gaps) needs to be part of RES for your particular use-case. If you want to publish events on positions 10-20, you should be able to observe an event on position 9 first. So you can query for that before doing publish. And if it is not there, you can raise an exception and retry later. The effect would be identical to what #478 tries to achieve. The only difference is probably that you would be doing it 0.5ms earlier, but I suppose that's not an issue. What do you think?

p.s. What's the order of magnitude of the rate of incoming messages per second? 1, 10, 1000?

**Comment by pawelpacana** [comment]
@vaihtovirta Clarification on this behavior in docs would be appreciated. Would you like to make a PR? 

That page is https://github.com/RailsEventStore/rails_event_store/blob/master/railseventstore.org/source/docs/expected_version.html.md

I'm going to withhold with [preliminary check](https://github.com/RailsEventStore/rails_event_store/pull/478) for now but I'll gladly accept a PR with optional check moved to repositories (instead of `ExpectedVersion` class). 
Btw. it seems that https://github.com/RailsEventStore/rails_event_store/blob/2a06ee97dabba7a7683be55d0a4343b9b1311f65/ruby_event_store/lib/ruby_event_store/in_memory_repository.rb#L111 is already strict in that regard by default.

**Comment by paneq** [comment]
> Btw. it seems that https://github.com/RailsEventStore/rails_event_store/blob/2a06ee97dabba7a7683be55d0a4343b9b1311f65/ruby_event_store/lib/ruby_event_store/in_memory_repository.rb#L111 is already strict in that regard by default.

I think we (maybe me, I don't remember exactly) added it to catch up such potential issues in tests before they creep into production.
